### PR TITLE
[0.17] Add OptionsCategory::ELEMENTS

### PR DESCRIPTION
**Pull request by stevenroose**
Attempt to group elements-specific -options together.

### CODE REVIEW

**Comment by instagibbs** [comment]
utACK

**Comment by stevenroose** [comment]
Not sure how this category should be applied to the CHAINPARAMS category in chainparams.cpp. Those are also mostly elements-specific. (fedpegscript, blocksignscript, ...)

**Comment by instagibbs** [comment]
I say let's add all of them if we're bothering to do any of them, to make it easier to rebase later.

File: src/init.cpp [hunk 1]
```diff
@@ -521,8 +520,23 @@ void SetupServerArgs()
     hidden_args.emplace_back("-daemon");
 #endif
 
+    //
+    // Elements-specific arguments.
+    //
+
+    std::vector<std::string> elements_hidden_args = {"-con_fpowallowmindifficultyblocks", "-con_fpownoretargeting", "-con_nsubsidyhalvinginterval", "-con_bip16exception", "-con_bip34height", "-con_bip65height", "-con_bip66height", "-con_npowtargettimespan", "-con_npowtargetspacing", "-con_nrulechangeactivationthreshold", "-con_nminerconfirmationwindow", "-con_powlimit", "-con_bip34hash", "-con_nminimumchainwork", "-con_defaultassumevalid", "-npruneafterheight", "-fdefaultconsistencychecks", "-fmineblocksondemand", "-fallback_fee_enabled", "-pchmessagestart"};
```
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/init.cpp]: What's the exact difference between hidden and debug? Actually I don't really recall where I got this list. I think they were all part of @jtimon's testchains PR (https://github.com/ElementsProject/elements/pull/432).

File: src/init.cpp [hunk 2]
```diff
@@ -521,8 +520,23 @@ void SetupServerArgs()
...
+
+    gArgs.AddArg("-pubkeyprefix", strprintf("The byte prefix, in decimal, of the chain's base58 pubkey address. (default: %d)", 235), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-scriptprefix", strprintf("The byte prefix, in decimal, of the chain's base58 script address. (default: %d)", 75), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-secretprefix", strprintf("The byte prefix, in decimal, of the chain's base58 secret key encoding. (default: %d)", 239), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-extpubkeyprefix", strprintf("The 4-byte prefix, in hex, of the chain's base58 extended public key encoding. (default: %s)", "043587CF"), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-extprvkeyprefix", strprintf("The 4-byte prefix, in hex, of the chain's base58 extended private key encoding. (default: %s)", "04358394"), false, OptionsCategory::ELEMENTS);
+    gArgs.AddArg("-bech32_hrp", strprintf("The human-readable part of the chain's bech32 encoding. (default: %s)", "bc"), false, OptionsCategory::ELEMENTS);
```
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/init.cpp]: Hmm. Yeah I copied this from elements-0.14.1, but you're right.

[the remark this answers is not in the document]
Comment (stevenroose) [reply src/init.cpp]: This should be fixed using the `defaultChainParams`.

File: src/init.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtimon) [reply src/init.cpp]: Adding them as debug args sounds good to me. Perhaps do the same in https://github.com/bitcoin/bitcoin/pull/8994 ?

File: src/init.cpp [hunk 3]
```diff
@@ -521,8 +521,23 @@ void SetupServerArgs()
     hidden_args.emplace_back("-daemon");
 #endif
 
+    //
+    // Elements-specific arguments.
+    //
+
+    std::vector<std::string> elements_hidden_args = {"-con_fpowallowmindifficultyblocks", "-con_fpownoretargeting", "-con_nsubsidyhalvinginterval", "-con_bip16exception", "-con_bip34height", "-con_bip65height", "-con_bip66height", "-con_npowtargettimespan", "-con_npowtargetspacing", "-con_nrulechangeactivationthreshold", "-con_nminerconfirmationwindow", "-con_powlimit", "-con_bip34hash", "-con_nminimumchainwork", "-con_defaultassumevalid", "-npruneafterheight", "-fdefaultconsistencychecks", "-fmineblocksondemand", "-fallback_fee_enabled", "-pchmessagestart"};
+
+    gArgs.AddArg("-pubkeyprefix", strprintf("The byte prefix, in decimal, of the chain's base58 pubkey address. (default: %d)", defaultChainParams->Base58Prefix(CChainParams::PUBKEY_ADDRESS)[0]), false, OptionsCategory::ELEMENTS);
```
Comment (jtimon) [inline src/init.cpp]: These look more chainparams specific than elements specific to me.

**Comment by jtimon** [comment]
concept ACK (at least for any new parameter not included in https://github.com/bitcoin/bitcoin/pull/8994 ).

**Comment by stevenroose** [comment]
I updated the chainparams ones. For the debug ones from testchains, I'm waiting to know if I can inherit the ones from upstream.

File: src/init.cpp [hunk 4]
```diff
@@ -515,14 +515,30 @@ void SetupServerArgs()
     gArgs.AddArg("-rpcworkqueue=<n>", strprintf("Set the depth of the work queue to service RPC calls (default: %d)", DEFAULT_HTTP_WORKQUEUE), true, OptionsCategory::RPC);
     gArgs.AddArg("-server", "Accept command line and JSON-RPC commands", false, OptionsCategory::RPC);
 
+    // chain params
+    gArgs.AddArg("-pubkeyprefix", strprintf("The byte prefix, in decimal, of the chain's base58 pubkey address. (default: %d)", defaultChainParams->Base58Prefix(CChainParams::PUBKEY_ADDRESS)[0]), false, OptionsCategory::CHAINPARAMS);
```
Comment (instagibbs) [inline src/init.cpp]: We'll just have to make sure to update `defaultChainParams` to using Custom Chains instead of MAIN at some point.

**Comment by instagibbs** [comment]
ACK

**Pull request merged**